File: app/src/main/java/com/mkulesh/onpc/iscp/messages/PresetCommandMsg.java
/*
 * Enhanced Music Controller
 * Copyright (C) 2018-2024 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */

package com.mkulesh.onpc.iscp.messages;

import android.annotation.SuppressLint;

import com.mkulesh.onpc.R;
import com.mkulesh.onpc.iscp.EISCPMessage;
import com.mkulesh.onpc.iscp.ZonedMessage;
import com.mkulesh.onpc.utils.Logging;

import java.util.ArrayList;
import java.util.Collections;

import androidx.annotation.DrawableRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

/*
 * Preset Command (Include Tuner Pack Model Only)
 */
public class PresetCommandMsg extends ZonedMessage
{
    public final static String CODE = "PRS";
    final static String ZONE2_CODE = "PRZ";
    final static String ZONE3_CODE = "PR3";
    final static String ZONE4_CODE = "PR4";

    public final static String[] ZONE_COMMANDS = new String[]{ CODE, ZONE2_CODE, ZONE3_CODE, ZONE4_CODE };

    public final static int NO_PRESET = -1;

    public enum Command implements DcpStringParameterIf
    {
        UP(R.string.preset_command_up, R.drawable.cmd_right),
        DOWN(R.string.preset_command_down, R.drawable.cmd_left);

        @StringRes
        final int descriptionId;

        @DrawableRes
        final int imageId;

        Command(@StringRes final int descriptionId, @DrawableRes final int imageId)
        {
            this.descriptionId = descriptionId;
            this.imageId = imageId;
        }

        public String getCode()
        {
            return toString();
        }

        @NonNull
        public String getDcpCode()
        {
            return toString();
        }

        @StringRes
        public int getDescriptionId()
        {
            return descriptionId;
        }

        @DrawableRes
        public int getImageId()
        {
            return imageId;
        }
    }

    private final Command command;
    private final ReceiverInformationMsg.Preset presetConfig;
    private int preset = NO_PRESET;

    PresetCommandMsg(EISCPMessage raw) throws Exception
    {
        super(raw, ZONE_COMMANDS);
        command = (Command) searchParameter(data, Command.values(), null);
        presetConfig = null;
        try
        {
            preset = command == null ? Integer.parseInt(data, 16) : NO_PRESET;
        }
        catch (Exception e)
        {
            // nothing to do
        }
    }

    public PresetCommandMsg(int zoneIndex, final String command)
    {
        super(ZONE_COMMANDS, zoneIndex, command);
        this.command = (Command) searchParameter(command, Command.values(), null);
        this.presetConfig = null;
        this.preset = NO_PRESET;
    }

    public PresetCommandMsg(int zoneIndex, final ReceiverInformationMsg.Preset presetConfig, final int preset)
    {
        super(0, null, zoneIndex);
        this.command = null;
        this.presetConfig = presetConfig;
        this.preset = preset;
    }

    @Override
    public String getZoneCommand()
    {
        return ZONE_COMMANDS[zoneIndex];
    }

    public Command getCommand()
    {
        return command;
    }

    public ReceiverInformationMsg.Preset getPresetConfig()
    {
        return presetConfig;
    }

    public int getPreset()
    {
        return preset;
    }

    @NonNull
    @Override
    public String toString()
    {
        return getZoneCommand() + "[" + data
                + "; ZONE_INDEX=" + zoneIndex
                + "; CMD=" + (command != null ? command.toString() : "null")
                + "; PRS_CFG=" + (presetConfig != null ? presetConfig.getName() : "null")
                + "; PRESET=" + preset + "]";
    }

    @Override
    public EISCPMessage getCmdMsg()
    {
        String par = "";
        if (command != null)
        {
            par = command.getCode();
        }
        else if (presetConfig != null)
        {
            par = String.format("%02x", presetConfig.getId());
        }
        return new EISCPMessage(getZoneCommand(), par);
    }

    @Override
    public boolean hasImpactOnMediaList()
    {
        return false;
    }

    /*
     * Denon control protocol. The same message is used for both zones.
     */
    private final static String DCP_COMMAND = "TPAN";
    private final static String DCP_COMMAND_OFF = "OFF";

    @NonNull
    public static ArrayList<String> getAcceptedDcpCodes()
    {
        return new ArrayList<>(Collections.singletonList(DCP_COMMAND));
    }

    public static PresetCommandMsg processDcpMessage(@NonNull String dcpMsg, int zone)
    {
        if (dcpMsg.startsWith(DCP_COMMAND))
        {
            final String par = dcpMsg.substring(DCP_COMMAND.length()).trim();
            if (par.equalsIgnoreCase(DCP_COMMAND_OFF))
            {
                return new PresetCommandMsg(zone, null, NO_PRESET);
            }
            try
            {
                final int preset = Integer.parseInt(par);
                return new PresetCommandMsg(zone, null, preset);
            }
            catch (Exception e)
            {
                Logging.info(PresetCommandMsg.class, "Unable to parse preset " + par);
                return null;
            }
        }
        return null;
    }

    @SuppressLint("DefaultLocale")
    @Nullable
    @Override
    public String buildDcpMsg(boolean isQuery)
    {
        return DCP_COMMAND + (isQuery ? DCP_MSG_REQ :
                (command != null ? command.getDcpCode() : String.format("%02d", preset)));
    }

}

File: lib/iscp/messages/PresetCommandMsg.dart
/*
 * Enhanced Music Controller
 * Copyright (C) 2019-2023 by Mikhail Kulesh
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
 * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program.
 */
// @dart=2.9
import 'package:sprintf/sprintf.dart';

import "../../constants/Drawables.dart";
import "../../constants/Strings.dart";
import "../../utils/Logging.dart";
import "../EISCPMessage.dart";
import "../ISCPMessage.dart";
import "EnumParameterMsg.dart";
import "ReceiverInformationMsg.dart";

enum PresetCommand
{
    NONE,
    UP,
    DOWN
}

/*
 * Preset Command (Include Tuner Pack Model Only)
 */
class PresetCommandMsg extends ZonedMessage
{
    static const String CODE = "PRS";
    static const String ZONE2_CODE = "PRZ";
    static const String ZONE3_CODE = "PR3";
    static const String ZONE4_CODE = "PR4";

    static const List<String> ZONE_COMMANDS = [CODE, ZONE2_CODE, ZONE3_CODE, ZONE4_CODE];

    static const int NO_PRESET = -1;

    static const ExtEnum<PresetCommand> ValueEnum = ExtEnum<PresetCommand>([
        EnumItem(PresetCommand.NONE, defValue: true),
        EnumItem(PresetCommand.UP, dcpCode: "UP",
            descrList: Strings.l_preset_command_up, icon: Drawables.cmd_right),
        EnumItem(PresetCommand.DOWN, dcpCode: "DOWN",
            descrList: Strings.l_preset_command_down, icon: Drawables.cmd_left)
    ]);

    EnumItem<PresetCommand> _command;
    Preset _presetConfig;
    int _preset;

    PresetCommandMsg(EISCPMessage raw) : super(ZONE_COMMANDS, raw)
    {
        final EnumItem<PresetCommand> c = ValueEnum.valueByCode(getData);
        _command = c.key == PresetCommand.NONE ? null : c;
        _presetConfig = null;
        _preset =  _command == null?
            ISCPMessage.nonNullInteger(getData, 16, NO_PRESET) : NO_PRESET;
    }

    PresetCommandMsg.outputCmd(int zoneIndex, final PresetCommand command) :
            super.output(ZONE_COMMANDS, zoneIndex, ValueEnum.valueByKey(command).getCode)
    {
        _command = ValueEnum.valueByKey(command);
        _presetConfig = null;
        _preset = NO_PRESET;
    }

    PresetCommandMsg.outputCfg(int zoneIndex, final Preset presetConfig) :
            super.output(ZONE_COMMANDS, zoneIndex, _getParameterAsString(presetConfig))
    {
        _command = null;
        _presetConfig = presetConfig;
        _preset = NO_PRESET;
    }

    PresetCommandMsg.dcp(int zoneIndex, int preset) :
            super.output(ZONE_COMMANDS, zoneIndex, preset.toString())
    {
        _command = null;
        _presetConfig = null;
        _preset = preset;
    }

    static String _getParameterAsString(final Preset presetConfig)
    {
        return presetConfig.getId.toRadixString(16).padLeft(2, '0');
    }

    EnumItem<PresetCommand> get getCommand
    => _command;

    Preset get getPresetConfig
    => _presetConfig;

    int get getPreset
    => _preset;

    @override
    String toString()
    => super.toString()
            + "[CMD=" + (_command != null ? _command.toString() : "null")
            + "; PRS_CFG=" + (_presetConfig != null ? _presetConfig.getName : "null")
            + "; PRESET=" + _preset.toString() + "]";

    @override
    bool hasImpactOnMediaList()
    {
        return false;
    }

    /*
     * Denon control protocol. The same message is used for both zones.
     */
    static const String _DCP_COMMAND = "TPAN";
    static const String _DCP_COMMAND_OFF = "OFF";

    static List<String> getAcceptedDcpCodes()
    => [ _DCP_COMMAND ];

    static PresetCommandMsg processDcpMessage(String dcpMsg, int zone)
    {
        if (dcpMsg.startsWith(_DCP_COMMAND))
        {
            final String par = dcpMsg.substring(_DCP_COMMAND.length).trim();
            if (par == _DCP_COMMAND_OFF)
            {
                return PresetCommandMsg.dcp(zone, NO_PRESET);
            }
            if (int.tryParse(par) != null)
            {
                final int preset = int.tryParse(par);
                return PresetCommandMsg.dcp(zone, preset);
            }
            else
            {
                Logging.info(PresetCommandMsg, "Unable to parse preset " + par);
                return null;
            }
        }
        return null;
    }

    @override
    String buildDcpMsg(bool isQuery)
    => _DCP_COMMAND + (isQuery ? ISCPMessage.DCP_MSG_REQ :
        (_command != null ? _command.getDcpCode : sprintf("%02d", [ _preset ])));
}
